File: server/core/run_energy.py
"""Level-running kcal from mass × distance; theoretical fat equivalent."""

from __future__ import annotations

KCAL_PER_KG_FAT_THEORY = 7700.0

# Empirical level running: commonly quoted ~0.9–1.1 kcal per kg body mass per km.
KCAL_PER_KG_KM_RUNNING = 1.0


def kcal_running_mass_distance(weight_kg: float, distance_km: float) -> float:
    return weight_kg * distance_km * KCAL_PER_KG_KM_RUNNING


def fat_equiv_kg_from_kcal(kcal: float) -> float:
    return kcal / KCAL_PER_KG_FAT_THEORY


def distance_km_from_speed(speed_kmh: float, duration_seconds: float) -> float:
    return speed_kmh * (duration_seconds / 3600.0)

File: client/lib/run-energy.ts
/** Level-running kcal from mass × distance; mirrors `server/core/run_energy.py`. */

export const KCAL_PER_KG_FAT_THEORY = 7700;

/** ~1 kcal per kg per km on level ground (literature often ~0.9–1.1). */
export const KCAL_PER_KG_KM_RUNNING = 1.0;

export function kcalRunningMassDistance(
  weightKg: number,
  distanceKm: number,
  met: number = KCAL_PER_KG_KM_RUNNING
): number {
  return weightKg * distanceKm * met;
}

export function fatEquivKgFromKcal(kcal: number): number {
  return kcal / KCAL_PER_KG_FAT_THEORY;
}

export function distanceKmFromSpeed(
  speedKmh: number,
  durationSeconds: number
): number {
  return speedKmh * (durationSeconds / 3600);
}

/** Assumed average step length used to estimate step count from distance — no GPS/pedometer input exists, so steps are always derived. */
export const STEP_LENGTH_M = 0.75;

export function stepsFromDistanceKm(distanceKm: number): number {
  return Math.round((distanceKm * 1000) / STEP_LENGTH_M);
}
